pkg/kine/drivers/generic/batch.go

```diff
@@ -0,0 +1,206 @@
+package generic
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+
+	"github.com/canonical/k8s-dqlite/pkg/kine/prepared"
+)
+
+type BatchConn interface {
+	ExecContext(ctx context.Context, query string, args ...any) (result sql.Result, err error)
+}
+
+var _ BatchConn = &sql.DB{}
+var _ BatchConn = &sql.Tx{}
+var _ BatchConn = &sql.Conn{}
+
+var _ BatchConn = &prepared.DB{}
+var _ BatchConn = &prepared.Tx{}
+
+type batchStatus int
+
+const (
+	batchNotStarted batchStatus = iota
+	batchStarted
+	batchRunning
+)
+
+type Batch struct {
+	db     *prepared.DB
+	mu     sync.Mutex
+	cv     sync.Cond
+	status batchStatus
+
+	queue []*batchJob
+	runId int64
+}
+
+func NewBatch(db *prepared.DB) *Batch {
+	b := &Batch{
+		db: db,
+	}
+	b.cv.L = &b.mu
+	return b
+}
+
+func (b *Batch) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	runId := b.runId
+	if b.status == batchRunning {
+		// The current run is already taking place.
+		runId++
+	}
+
+	job := &batchJob{
+		ctx:   ctx,
+		query: query,
+		args:  args,
+		runId: runId,
+	}
+	b.queue = append(b.queue, job)
+
+	b.run()
+
+	for job.runId >= b.runId {
+		b.cv.Wait()
+	}
+
+	if job.err != nil {
+		return nil, job.err
+	}
+
+	return job, nil
+}
+
+// run starts a batching job which will run until queue exaustion.
+// run does not block other goroutine from enqueuing new jobs.
+//
+// It must be called while holding the mu lock.
+func (b *Batch) run() {
+	if b.status == batchNotStarted {
+		b.status = batchStarted
+
+		go func() {
+			b.mu.Lock()
+			defer b.mu.Unlock()
+
+			b.status = batchRunning
+			defer func() { b.status = batchNotStarted }()
+
+			for len(b.queue) > 0 {
+				queue := b.queue
+				b.queue = nil
```

Comment:
I may be missing something here, but it looks like this line makes the loop only run once? This feels either wrong or fragile to me. Once a batch has started executing, do we want new jobs to be added to its queue, allowing newer items to be merged into the current batch? Although perhaps more performant, this feels like it would make it harder to reason about rollbacks. Should the condition in the `for` be `len(queue) > 0`?